File: 2015/day3PerfectlySphericalHousesInAVacuum/day3.mjs
// --- Day 3: Perfectly Spherical Houses in a Vacuum ---

import { readFileSync } from "node:fs";

const input = "day3Input.txt";

const directions = readFileSync(input, { encoding: "utf-8" }).split("");

const directionsMechanism = {
  "<": { axis: "x", value: -1 },
  ">": { axis: "x", value: 1 },
  "^": { axis: "y", value: 1 },
  v: { axis: "y", value: -1 },
};

const coordinates = { x: 0, y: 0 };

let housesVisited = { "0,0": 1 };

directions.forEach((direction) => {
  const { axis, value } = directionsMechanism[direction];

  coordinates[axis] += value;

  const currentPosition = `${coordinates.x},${coordinates.y}`;
  housesVisited[currentPosition] = housesVisited[currentPosition]
    ? housesVisited[currentPosition] + 1
    : 1;
});

const numHousesWithPresents = Object.keys(housesVisited).length;
console.log(
  `The number of houses that receive at least one present is: ${numHousesWithPresents}`
); // Answer Part 1

// Part 2

const coordinatesSanta = { x: 0, y: 0 };

const coordinatesRoboSanta = { x: 0, y: 0 };

housesVisited = { "0,0": 2 };

directions.forEach((direction, index) => {
  const turn = index + 1;
  const { axis, value } = directionsMechanism[direction];

  let newPosition;

  if (turn % 2 === 0) {
    coordinatesRoboSanta[axis] += value;
    newPosition = `${coordinatesRoboSanta.x},${coordinatesRoboSanta.y}`;
  } else {
    coordinatesSanta[axis] += value;
    newPosition = `${coordinatesSanta.x},${coordinatesSanta.y}`;
  }

  housesVisited[newPosition] = housesVisited[newPosition]
    ? housesVisited[newPosition] + 1
    : 1;
});

const numHousesWithPresentsNextYear = Object.keys(housesVisited).length;

console.log(
  `The number of houses that receive at least one present on the next year is: ${numHousesWithPresentsNextYear}`
); // Answer Part 2

File: 2015/day3PerfectlySphericalHousesInAVacuum/day3.py
# --- Day 3: Perfectly Spherical Houses in a Vacuum ---

input = "day3Input.txt"

with open(input, "r", encoding="utf-8") as file:
    directions = file.read()

directions_mechanism = {
    "<": {"axis": "x", "value": -1},
    ">": {"axis": "x", "value": 1},
    "^": {"axis": "y", "value": 1},
    "v": {"axis": "y", "value": -1},
}

coordinates = {"x": 0, "y": 0}

houses_visited = {(0, 0): 1}

for direction in directions:
    axis = directions_mechanism[direction]["axis"]
    value = directions_mechanism[direction]["value"]
    coordinates[axis] += value

    current_position = (coordinates["x"], coordinates["y"])
    houses_visited[current_position] = houses_visited.get(current_position, 0) + 1

num_houses_with_presents = len(houses_visited)
print(
    f"The number of houses that receive at least one present is: {num_houses_with_presents}"
) # Answer Part 1
